src/sinks/redis.rs

```diff
@@ -0,0 +1,587 @@
+) -> Option<EncodedEvent<RedisKvEntry>> {
+    let key = key
+        .render_string(&event)
+        .map_err(|error| {
+            emit!(TemplateRenderingFailed {
+                error,
+                field: Some("key"),
+                drop_event: true,
+            });
+        })
+        .ok()?;
+
+    encoding.apply_rules(&mut event);
+
+    let value = match encoding.codec() {
+        Encoding::Json => serde_json::to_vec(event.as_log())
+            .map_err(|error| panic!("Unable to encode into JSON: {}", error))
+            .unwrap_or_default(),
+        Encoding::Text => event
+            .as_log()
+            .get(log_schema().message_key())
+            .map(|v| v.as_bytes().to_vec())
+            .unwrap_or_default(),
+    };
+
+    let event = EncodedEvent::new(RedisKvEntry { key, value });
+    Some(event)
+}
+
+type RedisPipeResult = RedisResult<Vec<bool>>;
+
+impl Response for Vec<bool> {
+    fn is_successful(&self) -> bool {
+        let mut flag = true;
+        for r in self {
+            if !r {
+                flag = false;
+                break;
+            }
+        }
+        flag
+    }
+}
+
+#[derive(Debug, Clone)]
+struct RedisRetryLogic;
+
+impl RetryLogic for RedisRetryLogic {
+    type Error = RedisError;
+    type Response = Vec<bool>;
+
+    fn is_retriable_error(&self, _error: &Self::Error) -> bool {
+        true
+    }
+
+    fn should_retry_response(&self, response: &Self::Response) -> RetryAction {
+        if response.is_successful() {
+            return RetryAction::Successful;
+        }
+        RetryAction::Retry("Sent data to redis failed.".into())
+    }
+}
+
+#[derive(Clone)]
+pub struct RedisSink {
+    conn: ConnectionManager,
+    data_type: DataType,
+    method: Option<Method>,
+}
+
+impl Service<Vec<RedisKvEntry>> for RedisSink {
+    type Response = Vec<bool>;
+    type Error = RedisError;
+    type Future = BoxFuture<'static, RedisPipeResult>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, kvs: Vec<RedisKvEntry>) -> Self::Future {
+        let mut byte_size = 0;
+
+        let data_type = self.data_type;
+        let method = self.method;
+
+        let mut conn = self.conn.clone();
+        let mut pipe = redis::pipe();
+
+        for kv in kvs {
+            byte_size += kv.encoded_length();
+            match (data_type, method) {
+                (DataType::List, Some(Method::LPush)) => {
+                    pipe.atomic().lpush(kv.key, kv.value);
+                }
+                (DataType::List, Some(Method::RPush)) => {
+                    pipe.atomic().rpush(kv.key, kv.value);
+                }
+                (DataType::Channel, None) => {
+                    pipe.atomic().publish(kv.key, kv.value);
+                }
+                _ => {
+                    panic!("The `data_type` can't be empty, when `data_type` is `list`, `method` cannot be empty.")
```

Comment:
I think we could model this using an enum to wrap the data type + method so this `match` wouldn't need to check for invalid variants.

Maybe something like:

```
enum DataType {
  List(Method)
  Channel
}
```

The existing `DataType` could be renamed to `DataTypeConfig`.

Currently, it seems like it will error at runtime if `data_type` is `list` and `method` is missing. Ideally we'd catch that at vector start-up time.